### PR TITLE
feat: add `lwc:on` directive

### PR DESCRIPTION
**Pull request by gaurav-rk9**
## Details

Implement `lwc:on` as specified in [RFC: lwc:on Directive](https://github.com/salesforce/lwc-rfcs/pull/92)

differences from RFC : 

- When `lwc:on={eventHandlers}` is used, the object passed, i.e. `eventHandlers` will be freezed, and any attempt to modify it shall throw a runtime error. However `eventHandlers` can be re-assigned to a different object, and this shall update the actual event listeners correspondingly, i.e. removal of previous event listeners and addition of new ones.
- disallows usage of `onXXX` and `lwc:on` together on HTML templates.

## Does this pull request introduce a breaking change?

- 😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

- 🔬 Yes, it does include an observable change.

Allows the use of new `lwc:on` directive

## GUS work item

W-17411303, W-17632714

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/api.ts [hunk 1]
```diff
@@ -716,6 +716,11 @@ function shc(content: unknown): SanitizedHtmlContent {
     return createSanitizedHtmlContent(sanitizedString);
 }
 
+// [fr]ee[z]e
+function frz(obj: object) {
+    return ObjectFreeze(obj);
```
Comment (gaurav-rk9) [inline packages/@lwc/engine-core/src/framework/api.ts:721]: This doesn't work when the object is decorated with @api
Also need to improve this to be less restrictive and possibly throw custom error.

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
```
Comment (divmain) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:2]:
```suggestion
 * Copyright (c) 2025, salesforce.com, inc.
```

File: packages/@lwc/engine-core/src/framework/vm.ts [hunk 3]
```diff
@@ -135,6 +135,8 @@ export interface VM<N = HostNode, E = HostElement> {
     readonly owner: VM<N, E> | null;
     /** References to elements rendered using lwc:ref (template refs) */
     refVNodes: RefVNodes | null;
+    /** Actual listeners added to elements corresponding to listeners provided by lwc:on */
+    actualEventListeners: WeakMap<Element, Record<string, EventListener>>;
```
Comment (wjhsf) [inline packages/@lwc/engine-core/src/framework/vm.ts:139]: Why `actual` event listeners?

File: packages/@lwc/template-compiler/src/parser/index.ts [hunk 4]
```diff
@@ -972,6 +973,31 @@ function applyLwcSpreadDirective(
...
+    const lwcOn = parsedAttr.pick(ElementDirectiveName.On);
+    if (!lwcOn) {
+        return;
+    }
+
+    const { value: lwcOnValue } = lwcOn;
+    if (!ast.isExpression(lwcOnValue)) {
+        ctx.throwOnNode(ParserDiagnostics.INVALID_LWC_ON_LITERAL_PROP, element, [`<${tag}>`]);
+    }
+    if (element.listeners.length) {
```
Comment (wjhsf) [inline packages/@lwc/template-compiler/src/parser/index.ts:992]: Please add a comment explaining why we are doing this check.

File: packages/@lwc/template-compiler/src/codegen/index.ts [hunk 5]
```diff
@@ -627,6 +629,10 @@ function transform(codeGen: CodeGen): t.Expression {
             data.push(codeGen.genEventListeners(listeners));
         }
 
+        if (onDirective) {
+            data.push(codeGen.genDynamicEventListeners(onDirective));
```
Comment (wjhsf) [inline packages/@lwc/template-compiler/src/codegen/index.ts:633]: Can you add a comment explaining what is happening here?

File: packages/@lwc/template-compiler/src/codegen/codegen.ts [hunk 6]
```diff
@@ -456,6 +459,12 @@ export default class CodeGen {
         }
     }
 
+    genDynamicEventListeners(onDirective: OnDirective) {
+        const onDirectiveArg = this.bindExpression(onDirective.value);
+        const freezedArg = this._renderApiCall(RENDER_APIS.freeze, [onDirectiveArg]);
+        return t.property(t.identifier('dynamicOn'), freezedArg);
```
Comment (wjhsf) [inline packages/@lwc/template-compiler/src/codegen/codegen.ts:465]: Indirect AST building is hard to parse. Can you add a comment with the expected result?

File: packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { LightningElement } from 'lwc';
+
+export default class Basic extends LightningElement {
+    privateVariable = "'this' is the component"; // available only on the component not on the element
+
+    /* eslint-disable no-console */
```
Comment (wjhsf) [inline packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js:6]: Instead of doing `console.log` in every test file, can you do `@api callback` and use that callback? That way, when you're developing this, you can set `callback` to `console.log`, but we can use a no-op by default to avoid too much noise in the console.

File: packages/@lwc/integration-karma/test/lwc-on/index.spec.js [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        const element = createElement('x-lifecycle', { is: Lifecycle });
+
+        document.body.appendChild(element);
+
+        expect(consoleSpy).toHaveBeenCalledWith(
+            'handled events dispatched from child connectedCallback'
+        );
+    });
+
+    // describe('re-render behavior', () => {
```
Comment (wjhsf) [inline packages/@lwc/integration-karma/test/lwc-on/index.spec.js:146]: Please let us know when this is ready!

File: packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import { LightningElement } from 'lwc';
+
+export default class CaseVariants extends LightningElement {
+    /* eslint-disable no-console */
+    eventHandlers = {
```
Comment (wjhsf) [inline packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js:5]: Can we add a test case for a computed key? Something like `['computed']`, just to see what happens.

File: packages/@lwc/integration-karma/test/lwc-on/index.spec.js [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+        it('event listeners are added independently from lwc:on and lwc:spread', () => {
+            const element = createElement('x-spread', { is: Spread });
+            document.body.appendChild(element);
+
+            const button = element.shadowRoot.querySelector('button');
+            button.click();
+
+            expect(consoleSpy).toHaveBeenCalledWith('lwc:spread handler called');
+            expect(consoleSpy).toHaveBeenCalledWith('lwc:on handler called');
+        });
```
Comment (divmain) [inline packages/@lwc/integration-karma/test/lwc-on/index.spec.js:133]: An error should be thrown in this case. Just as we will not permit `onclick` alongside `lwc:on` for the same element, we cannot permit `onclick` being attached via `lwc:spread`. This cannot be caught at compile time (because we won't necessarily know what value the `lwc:spread` will receive. But it should result in a runtime error the component author attempts to attach event handliers via both `lwc:spread` and `lwc:on`.

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        data: { dynamicOn },
+    } = vnode;
+
+    // The argument passed to lwc:on is frozen.
+    // Because of this, `oldVnode.data.dynamicOn` is deeply equal
+    // to its value from the previous render cycle.
+    const oldDynamicOn =
+        isNull(oldVnode) || isUndefined(oldVnode.data.dynamicOn)
+            ? EmptyObject
+            : oldVnode.data.dynamicOn;
```
Comment (wjhsf) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:31]:
```suggestion
        oldVnode?.data?.dynamicOn ?? EmptyObject;
```

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    const newDynamicOn = isUndefined(dynamicOn) ? EmptyObject : dynamicOn;
```
Comment (wjhsf) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:32]:
```suggestion
    const newDynamicOn = dynamicOn ?? EmptyObject;
```

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        return;
+    }
+
+    const { addEventListener, removeEventListener } = renderer;
+    const actualEventListeners = getActualEventListeners(owner, elm!);
+
+    const oldDynamicOnNames = keys(oldDynamicOn);
+    const newDynamicOnNames = keys(newDynamicOn);
+
+    for (const name of oldDynamicOnNames) {
```
Comment (divmain) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:44]:
```suggestion
    // Remove listeners that were attached previously but don't have a corresponding listener in newDynamicOn
    for (const name of oldDynamicOnNames) {
```

Comment (wjhsf) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:44]: There are a lot of very similar names flying around here, can you add comments clarifying the intent of each `for`/`if` block?

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        if (!propertyIsEnumerable.call(newDynamicOn, name)) {
+            const actualListener = actualEventListeners[name];
+            removeEventListener(elm, name, actualListener);
+            delete actualEventListeners[name];
+        }
+    }
+
+    for (const name of newDynamicOnNames) {
```
Comment (divmain) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:52]:
```suggestion
    // Ensure that the event listeners that are attached match what is present in `newDynamicOnNames`
    for (const name of newDynamicOnNames) {
```

File: packages/@lwc/engine-core/src/framework/api.ts (hunk 1 above)
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/api.ts]: The reason this doesn't work with @api decorator is that the object recieved via public property is not the actual object but a read-only proxy of the actual object which doesn't allow any mutation

Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/api.ts]: In general, I think that framework should not mutate any user-land owned objects, doing so will be a big source of confusion

Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/api.ts]: I have updated the api completely, rather than freezing the actual object, we make a copy of the object and use that to validate for mutations.
This changes the timing of the error and has other functional  differences

File: packages/@lwc/integration-karma/test/lwc-on/index.spec.js (hunk 10 above)
Comment (gaurav-rk9) [reply packages/@lwc/integration-karma/test/lwc-on/index.spec.js]: I don't think we should do this. `onclick` added by `lwc:spread` has its equivalent in JS `element.onclick` , so throwing here would be a confusing behaviour.
Also the closest equivalent to this is `onclick={handler} lwc:spread={sreadObj}` in template HTML where `spreadObj={onclick:otherHandler}` . In this case we apply both the listeners and do not throw. `lwc:on` being an extension of `onclick`, I think we should maintain the same.
In general, I think we should discourage people from using `lwc:spread` with `onclick` via documentation but we should not be failing if one uses that.

Alternatives :
- remove `element.onclick` entirely, this would automatically remove spread behaviour too.
- unifying `element.onclick` and `onclick` added by `onclick`/`lwc:on` in HTML template to mean the same thing.
Both of these alternatives are breaking changes.

Also see - https://github.com/salesforce/lwc-rfcs/pull/92#discussion_r1954878116 thread

File: packages/@lwc/engine-core/src/framework/vm.ts (hunk 3 above)
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/vm.ts]: used `actual` to distinguish it from the provided functions in lwc:on. 
I have renamed variables to be more meaningful

File: packages/@lwc/template-compiler/src/parser/index.ts (hunk 4 above)
Comment (gaurav-rk9) [reply packages/@lwc/template-compiler/src/parser/index.ts:995]: done

File: packages/@lwc/template-compiler/src/codegen/index.ts (hunk 5 above)
Comment (gaurav-rk9) [reply packages/@lwc/template-compiler/src/codegen/index.ts:635]: Done

File: packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js (hunk 7 above)
Comment (gaurav-rk9) [reply packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js:6]: I am using spy (https://github.com/salesforce/lwc/pull/5228/files#diff-d1da7428b2d50ae19bfce55c9bed01ab52fb115257a9f0b598fc58abc9c26b2cR14), so there won't be any noise in actual console.
I intentionally used a global function to reduce test component complexity

File: packages/@lwc/integration-karma/test/lwc-on/index.spec.js (hunk 8 above)
Comment (gaurav-rk9) [reply packages/@lwc/integration-karma/test/lwc-on/index.spec.js]: ready

File: packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js (hunk 9 above)
Comment (gaurav-rk9) [reply packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js]: I don't understand what would be the benefit of this test.
My understaning is when computed property syntax in used to create an object, once the object is created, the object doesn't have any memory of computing expression. It is indistinguishable from the object which is created directly.

File: packages/@lwc/template-compiler/src/codegen/codegen.ts [hunk 15]
```diff
@@ -456,6 +459,50 @@ export default class CodeGen {
         }
     }
 
+    genDynamicEventListeners(onDirective: OnDirective) {
```
Comment (gaurav-rk9) [inline packages/@lwc/template-compiler/src/codegen/codegen.ts:462]: this would break when used inside loops, since it will be using the same context variable for all iterations

Comment (gaurav-rk9) [reply packages/@lwc/template-compiler/src/codegen/codegen.ts:460]: modified

File: packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js (hunk 9 above)
Comment (wjhsf) [reply packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js]: Actually, to catch the potential issue I'm concerned about, it should be `[computed]` (identifier, not string literal). The LWC compiler walks the AST of the code to apply transformations, and we want to validate that the transformations are applied correctly. I want to make sure we avoid a scenario like #4988, where the input `{[computed]: x}` is transformed to `{computed: x}`, rather than `{[computed]: x}`.

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+    const { addEventListener, removeEventListener } = renderer;
+    const attachedEventListeners = getAttachedEventListeners(owner, elm!);
+
+    // Properties that are present in 'oldDynamicOn' but not in 'newDynamicOn'
+    for (const eventType in oldDynamicOn) {
+        if (!(eventType in newDynamicOn)) {
+            // Throw if same object is passed
+            if (isObjectSame) {
+                throw new Error(
```
Comment (wjhsf) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:40]: I feel like, historically, we've made things like this just console warnings that only throw in dev mode. I also feel like, currently, we've decided that's an annoying pattern that nobody pays attention to, so maybe we should just go with this. @divmain, thoughts?

File: packages/@lwc/template-compiler/src/codegen/codegen.ts [hunk 17]
```diff
@@ -456,6 +457,25 @@ export default class CodeGen {
...
+        // Example Input : lwc:on={someObj}
+
+        // $cmp.someObj
+        const rawValue = this.bindExpression(onDirective.value);
+
+        // {__proto__: null, ...$cmp.someObj}
+        const clonedValue = t.objectExpression([
+            t.property(t.identifier('__proto__'), t.literal(null)),
+            t.spreadElement(rawValue),
+        ]);
```
Comment (wjhsf) [inline packages/@lwc/template-compiler/src/codegen/codegen.ts:470]: `__proto__` is deprecated (and hacky) and shouldn't be used. Instead, do `Object.assign(Object.create(null), <whatever>)`.

File: packages/@lwc/integration-karma/test/lwc-on/index.spec.js (hunk 10 above)
Comment (divmain) [reply packages/@lwc/integration-karma/test/lwc-on/index.spec.js]: I'm not sure I entirely agree. But it isn't something that I see as terribly important and, because `lwc:on` is net-new feature, it isn't something that would break functionality in existing components (which is my biggest concern right now). So if you are willing to work with our doc writers to update LWC and Salesforce Dev docs with the correct guidance (e.g. "combining `lwc:spread` with `lwc:on` is not recommended), I'm willing to concede this issue.

File: packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js (hunk 7 above)
Comment (divmain) [reply packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js:6]: For consistency with our codebase, please either use `TestUtils.spyConsole` or switch to passing in a `testCallback` as property to the component under test.; this keeps things consistent with the rest of our codebase. It has generally be a long-standing desire to move away from console spies.

If you go the `testCallback` route, add the `@api onTestClick` and `@api onMouseOver` to the component, swap `console.log` for `this.*`, and then your test would look something like this:

```javascript
        const onTestClick = createSpy('click handler');
        const onTestMouse = createSpy('mouse handler');
        const element = createElement('x-basic', { is: Basic });
        element.onTestClick = onTestClick;
        element.onTestMouse = onTestMouse;
        document.body.appendChild(element);

        const button = element.shadowRoot.querySelector('button');
        button.click();
        button.dispatchEvent(new MouseEvent('mouseover'));

        expect(onTestClick).toHaveBeenCalledWith('click handler called');
        expect(onTestMouse).toHaveBeenCalledWith('mouseover handler called');
```

Yes, it is more verbose (bad) but less magic (good) and more consistent with the rest of the codebase (very good).

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+                    `Detected mutation of property '${eventType}' in the object passed to lwc:on for '${sel}'. Reusing the same object with modified properties is prohibited. Please pass a new object instead.`
+                );
+            }
+
+            // Remove listeners that were attached previously but don't have a corresponding property in `newDynamicOn`
+            const attachedEventListener = attachedEventListeners[eventType];
+            removeEventListener(elm, eventType, attachedEventListener);
+            delete attachedEventListeners[eventType];
```
Comment (divmain) [inline packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:48]: Typically we try to avoid use of `delete` since it is a common cause of JIT deoptimization. However, this code path will only be reached occasionally, and this would only become an issue if there were literally thousands of event listeners being detached. I don't find that to be very likely, so this is okay as-is. I'm leaving the comment here for posterity, in case it _does_ become an issue in the future.

**Review by divmain — APPROVED** [review]
Just a few minor changes requested, but otherwise looks really good! Thanks for your work on this!

File: packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js (hunk 7 above)
Comment (divmain) [reply packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js:6]: Final note: it is okay with me if you want to make these changes in a follow-up PR. We are all on deadlines this week, and the requested changes here in the test file would have no bearing on what's shipped in the LWC libraries.

**Comment by gaurav-rk9** [comment]
Do we need to add any flags to toggle the availability of this feature, or is it not required?

**Comment by divmain** [comment]
> Do we need to add any flags to toggle the availability of this feature, or is it not required?

Realistically, yes @gaurav-rk9. Our policy is to add flags (and internal gates) for substantial changes. This qualifies, even if it is a net-new functionality. Probably the best place to do this is in the template parser, so that `lwc:on` nodes are never recognized as something special. Then, the compiled template will not include the calls to the new runtime logic you've added.

**Comment by divmain** [comment]
One last thing: there probably needs to be a small change in the SSR compiler to handle `lwc-on` template AST nodes by simply not emitting anything. That can be done in a followup PR, however.

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 16 above)
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:40]: I didn't notice earlier that `logError` is used elsewhere, In that case I would want to change this to `logError` since we would want that (reactivity with same object, modified) to be available if need has been demonstrated.
Also, that case (same object passed to `lwc:on`) works just as fine, if we don't throw, so there is no cost of 'making it work'.

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    const newDynamicOn = dynamicOn ?? EmptyObject;
+
+    // Compare raw values to check if same object is passed to lwc:on
+    const isObjectSame = oldVnode?.data?.dynamicOnRaw === dynamicOnRaw;
+
+    const { addEventListener, removeEventListener } = renderer;
+    const attachedEventListeners = getAttachedEventListeners(owner, elm!);
+
+    // Properties that are present in 'oldDynamicOn' but not in 'newDynamicOn'
+    for (const eventType in oldDynamicOn) {
```
[the remark this answers is not in the document]
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:36]: While copying, we have used the spread operator, which only copies own properties , so properties from `className.prototype` won't be copied.
I have also added a test to verify we are not adding listeners corresponding to inherited properties

File: packages/@lwc/template-compiler/src/codegen/codegen.ts (hunk 17 above)
Comment (gaurav-rk9) [reply packages/@lwc/template-compiler/src/codegen/codegen.ts:470]: [Runtime Semantics: PropertyDefinitionEvaluation](https://tc39.es/ecma262/#sec-runtime-semantics-propertydefinitionevaluation) - See "PropertyDefinition : PropertyName : AssignmentExpression" section , step 3 and 7. 
This is different from the legacy [Object.prototype.__proto__](https://tc39.es/ecma262/#sec-object.prototype.__proto__)

File: packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js (hunk 9 above)
Comment (gaurav-rk9) [reply packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js]: No transformation is applied to the object itself by the template compiler. It only works with the identifier representing the object, not the object directly. The meaning of that identifier is determined elsewhere, and this is a common abstraction used for all attributes in the template that accept variables as arguments — it’s not specific to `lwc:on`. 
So, I think to catch any issues arising from "the meaning of that identifier" being incorrect, we should have separate tests that specifically verifies that.

File: packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js (hunk 7 above)
Comment (gaurav-rk9) [reply packages/@lwc/integration-karma/test/lwc-on/x/basic/basic.js]: made the changes

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 18 above)
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts]: I was unaware of this.
On limited reading, it seems that `obj[x] = undefined` is better than `delete obj[x]`, so I have made the change

**Comment by gaurav-rk9** [comment]
> One last thing: there probably needs to be a small change in the SSR compiler to handle lwc-on template AST nodes by simply not emitting anything. That can be done in a followup PR, however.
It may take some time to understand SSR compiler, so I will do it in a followup PR

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:36]: Also, this (only own properties being used) is consistent with `lwc:spread`

**Comment by wjhsf** [comment]
@divmain should we use API versioning to introduce this change? It's a good candidate to incentivize people to upgrade.

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 16 above)
Comment (divmain) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:40]: Yes feel free.

File: packages/@lwc/template-compiler/src/codegen/codegen.ts (hunk 17 above)
Comment (divmain) [reply packages/@lwc/template-compiler/src/codegen/codegen.ts:470]: Good call out, you're right. I think this can remain as-is.

File: packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js (hunk 9 above)
Comment (divmain) [reply packages/@lwc/integration-karma/test/lwc-on/x/caseVariants/caseVariants.js]: I think what Will wants here is a sanity check in case the implementation changes in the future. You're correct that the object assigned to `eventHandlers` will not be touched by the template compiler. But it isn't impossible to imagine changes to the JS compiler in the future to resolve some bug with `lwc:on` or an adjacent feature.

Even the most innocuous code can be problematic when enough component authors are depending on niche observable behavior. So adding a test case here (while it may seem unnecessary) is a prudent thing to do.

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (divmain) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:36]: That sounds reasonable. Thanks!

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 18 above)
Comment (divmain) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts]: So long as this doesn't impact logic elsewhere (if you're enumerating keys or something like that), yes, this is preferable. Thanks!

**Review by wjhsf — COMMENTED** [review]
People try a lot of weird things, can you add some tests for some expected edge cases? Please add a test with `lwc:on={eventHandlers}` for each of these possible `eventHandlers`:

* null
* undefined
* string
* function
* defined as a setter on the class (without a getter)
* object with a getter that throws
* object with non-function values
* `new (class extends LightningElement {})` (and validate that a `render` event doesn't call the component's `render` method)

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 16 above)
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts]: done

File: packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (gaurav-rk9) [reply packages/@lwc/engine-core/src/framework/modules/dynamic-events.ts:39]: done

**Comment by Templarian** [comment]
@divmain @gaurav-rk9 I'm remove our workaround for this in early 258. Ping me once this makes it into a LWC version (doesn't need to be in the platform). Thanks.

**Comment by divmain** [comment]
These changes were merged as part of #5344.
